{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TCGA Survival Analysis\n",
    "\n",
    "**This analysis is based on the new interface implemented in `interface.py` and the functions in `DSA_utils.py`.**\n",
    "\n",
    "In this notebook we analyze the effect of mutations on patients survival over all cancer types and all TCGA therapies.\n",
    "\n",
    "Two operation modes are supported:\n",
    "1. Analyze an abritrary list of mutations.\n",
    "2. Analyze all mutations of a given gene.\n",
    "\n",
    "In general, given a list of TCGA mutations, the code analyzes the effect on survival over all cancer types and therapies, and reports the significant cases. Logrank test and Cox hazard model test are used to estimate the statistical difference between two groups of patients. In addition, a bootstrapping procedure is used, which controls for group sizes mismatch.\n",
    "\n",
    "Usage:\n",
    "- Run the section `Initialization` (should be executed only once when performing multiple testing). Modify the parameters in this section if needed.\n",
    "- Set the parameters in the `Parameters` section.\n",
    "- Execute the section `Run` to run the test.\n",
    "- Optional: execute the section `Plot a Result` to plot the two groups survival curves of a result."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Description\n",
    "\n",
    "We perform differential survival (DS) analysis based on a given `mutation`. The DS analysis compares the survival between two groups of patients, over all `cancer` types and all `therapy` types in the TCGA data, and produces a p-value for the null hypothesis that the survival in the two groups is the same.\n",
    "\n",
    "A test configuration, which defines a  DS test, contains the two group configurations. Each group configuration is a three-letter sequence, indicating the inclusion (capital letter) of exclusion (lower-case letter) of the `mutation`, `cancer`, and `therapy`, respectively, in the group of patients. For example, `MCT` defines a group of patients with the `mutation`, of `cancer` type that undergo `therapy`, whereas `mCt` defines group of patients **without** the `mutation`, of `cancer` type that **did not** undergo `therapy`. Specifically, `c` (i.e., lower-case c) implies all patients regardless of cancer type (i.e. PAN-cancer), and `t` (i.e. lower-case t) implies patients that did not undergo `therapy`.\n",
    "\n",
    "The current supported tests are:\n",
    "1. `MCT-mCT`: A single test. group1 contains all `cancer` patients with the `mutation` that undergo `therapy`, whereas group2 contains all `cancer` patients **without** the `mutation` that undergo `therapy`\n",
    "1. `McT-mcT`: A single test. group1 contains all patients (PAN-cancer) with the `mutation` that undergo `therapy`, whereas group2 contains all patients (PAN-cancer) **without** the `mutation` that undergo `therapy`\n",
    "1. `MCT-MCt,mCT-mCt`: Two tests. In the first test, group1 contains all `cancer` patients with the `mutation` that undergo `therapy`, whereas group2 contains all `cancer` patients with the `mutation` that **did not** undergo `therapy`. In the second test, group1 contains all `cancer` patients **without** the `mutation` that undergo `therapy`, whereas group2 contains all `cancer` patients **without** the `mutation` that **did not** undergo `therapy`\n",
    "1. `MCT-mCT,MCt-mCt`: Two tests. In the first test, group1 contains all `cancer` patients with the `mutation` that undergo `therapy`, whereas group2 contains all `cancer` patients **without** the `mutation` that undergo `therapy`. In the second test, group1 contains all `cancer` patients with the `mutation` that **did not** undergo `therapy`, whereas group2 contains all `cancer` patients **without** the `mutation` that **did not** undergo `therapy`\n",
    "\n",
    "In general, more tests can easily be supported.\n",
    "\n",
    "When a configuration contains multiple tests (as in the third and forth tests above), one of the tests can be defined as the `main` test. A `main` test implies, for a given `mutation`, that the tests **following** the `main` test are not executed unless a statistically significant DS was found in at least one `cancer` and/or `therapy` type in the `main` test (for that given `mutation`), in which case only these cases will be evalutaed in the following tests. Thus, a `main` test implies that the order of the tests is important. \n",
    "\n",
    "A `main` test is defined by the parameter `Main_test_index`. \n",
    "For example, for the `MCT-MCt,mCT-mCt` test and `Main_test_index=1`, the `mCT-mCt` will not be executed if no significant result was found for `MCT-MCt` (for a given mutation).\n",
    "Set `Main_test_index = <number of tests>` to disable this behavior."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reports\n",
    "\n",
    "For each test, the analysis generates two report files containing the `mutation`, `cancer` and `therapy` values that resulted in a statistically significant DS (non-significant results are not reported). Each report file includes the two group sizes, p values, and AUCs.\n",
    "The report files are:\n",
    "- A pickle file containing a dictionary with all the analysis data.\n",
    "- An excel file containing a summary report, where each row corresponds to a `test configuration` with `mutation`, `cancer` and `therapy` values that caused a significant DS."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notes\n",
    "The analysis below reports and saves only the significant results. If the non-significant results are also required, do the following:\n",
    "- Define `P_val_threshold` to be large (e.g. `P_val_threshold = 1.0`)\n",
    "- Run the analysis. All the p-values will be reported and saved.\n",
    "- Manually select the results that you consider as insignificant.\n",
    "\n",
    "In case of multiple tests (with `Main_test_index` as the main test), most of the cases will now be considered as significant, thus all the corresponding cases in the tests following the main tests will be executed, implying that the run time will increase.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Global configuration.\n",
    "'''\n",
    "import pathlib\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "# make sure PYTHONPATH is set, containing the Utils folder.\n",
    "import mysys_init as si\n",
    "import myutils as myut\n",
    "\n",
    "from Utils import DSA_utils as ut\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Initialization\n",
    "\n",
    "Run this section only once, even when running multiple tests multiple times. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" \n",
    "TCGA Data\n",
    "\"\"\"\n",
    "from Utils import TCGA_interface as TCGAi\n",
    "import GDC_dataset_feb2021_utils as gdcu  # needed only TCGA_mutation_data_folder in this cell\n",
    "\n",
    "# TCGA clinical file\n",
    "TCGA_clinical_data_file: pathlib.Path = pathlib.Path(si.ROOT_path) / 'Data' / 'GDC' / 'Clinical' / 'df_p_all.pkl'\n",
    "\n",
    "# TCGA main mutation folder\n",
    "TCGA_mutation_data_folder: pathlib.Path = pathlib.Path(si.ROOT_path) / 'Data' / 'GDC' / 'Mutations' / 'data' / 'filtered_feb_2021' / 'AllGenes' if si.MACHINE == 'MAC' else pathlib.Path(gdcu.Dataset_main_folder)\n",
    "\n",
    "# TCGA therapy and drug clinical data dictionary\n",
    "min_drug_prevalence: int = 20\n",
    "TCGA_df_clinical_drugs_dict_file_name: pathlib.Path = pathlib.Path(si.ROOT_path) / 'Explore_kaplan_meier' / 'Reports' / 'TCGA_clinical_preprocess' / f'df_clinical_drugs_dict_{min_drug_prevalence}.pkl'\n",
    "\n",
    "# instantiate TCGA data\n",
    "Dsa_Mut = TCGAi.TCGA_Mutation_ID()\n",
    "Dsa_data = TCGAi.TCGA_data(TCGA_mutation_data_folder, TCGA_clinical_data_file, TCGA_df_clinical_drugs_dict_file_name)\n",
    "Dsa_compute = TCGAi.TCGA_compute()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "report_path = pathlib.Path(si.ROOT_path) / 'Explore_kaplan_meier' / 'Develop1' / 'Reports'\n",
    "results_path = report_path / 'DSA_results'\n",
    "\n",
    "# therapies to ignore when evaluating the different therapies/drugs in the data\n",
    "#Ignore_therapies: set = {Dsa_data.None_therapy_name, Dsa_data.All_name, Dsa_data.All_therapies_name}\n",
    "Ignore_therapies: set = {Dsa_data.None_therapy_name, Dsa_data.All_therapies_name}\n",
    "\n",
    "# cancer types to ignore. E.g. setting to {'TCGA_GBM', 'TCGA_OV'} will discard patients of these cancer types.\n",
    "Ignore_cancers: set = set()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clinical data\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_barcode</th>\n",
       "      <th>case_id</th>\n",
       "      <th>sample_barcode</th>\n",
       "      <th>sample_uuid</th>\n",
       "      <th>female</th>\n",
       "      <th>male</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>alive</th>\n",
       "      <th>...</th>\n",
       "      <th>primary_tumor</th>\n",
       "      <th>blood_derived_normal</th>\n",
       "      <th>solid_tissue_normal</th>\n",
       "      <th>Proj_name</th>\n",
       "      <th>metastatic</th>\n",
       "      <th>recurrent_tumor</th>\n",
       "      <th>bone_marrow_normal</th>\n",
       "      <th>buccal_cell_normal</th>\n",
       "      <th>tumor</th>\n",
       "      <th>additional_metastatic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tcga-or-a5ld</td>\n",
       "      <td>d885b720-e813-4828-a731-b3cdcbd20687</td>\n",
       "      <td>tcga-or-a5ld-10a</td>\n",
       "      <td>7f0378a8-e904-404b-91de-a70e73bda1b4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>19111.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tcga-or-a5jd</td>\n",
       "      <td>240dca72-76e0-4d57-978d-1242e6e7ed68</td>\n",
       "      <td>tcga-or-a5jd-10b</td>\n",
       "      <td>4fdc28a5-f64e-4d98-baac-befad27fb2a1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>white</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20989.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tcga-or-a5jm</td>\n",
       "      <td>fb54458d-c373-46c2-841e-82663e13efaa</td>\n",
       "      <td>tcga-or-a5jm-10b</td>\n",
       "      <td>a3c80a41-bce1-4f72-8195-f1c699957c5f</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>white</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9359.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tcga-or-a5jz</td>\n",
       "      <td>94cdac79-2b4e-483a-8f17-3e0c9be589d0</td>\n",
       "      <td>tcga-or-a5jz-10a</td>\n",
       "      <td>a9c6d6e4-ab79-4a75-bd63-891647d82501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>white</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22218.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tcga-ou-a5pi</td>\n",
       "      <td>a5be4e42-cb6c-4e67-8547-45dbd058bb95</td>\n",
       "      <td>tcga-ou-a5pi-10a</td>\n",
       "      <td>a37b1255-a17c-49e9-97f7-8525aee7438b</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>19492.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11136</th>\n",
       "      <td>tcga-vd-aa8q</td>\n",
       "      <td>1d00217f-28df-4d91-aeb9-0d9a83a907e6</td>\n",
       "      <td>tcga-vd-aa8q-10a</td>\n",
       "      <td>d57d0601-a8eb-471e-9055-109caea0ecb9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26285.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UVM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11137</th>\n",
       "      <td>tcga-v4-a9ej</td>\n",
       "      <td>d9ccba9b-6bf0-44d9-8097-a97ab747f0bd</td>\n",
       "      <td>tcga-v4-a9ej-10a</td>\n",
       "      <td>cef70e8e-da78-4eec-ab47-63a8d2be2cc4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>13958.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UVM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11138</th>\n",
       "      <td>tcga-wc-a87u</td>\n",
       "      <td>4498044c-1607-403e-867e-befd7fad275a</td>\n",
       "      <td>tcga-wc-a87u-10a</td>\n",
       "      <td>929e6183-8e31-44ee-a2d9-89b3e8861e01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>26933.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UVM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11139</th>\n",
       "      <td>tcga-v4-a9eh</td>\n",
       "      <td>5e54122f-1c4d-45cc-8842-cd1c1df0eed8</td>\n",
       "      <td>tcga-v4-a9eh-10a</td>\n",
       "      <td>9256ff7e-5674-4e61-9fdc-947669e525c7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>white</td>\n",
       "      <td>hispanic or latino</td>\n",
       "      <td>19573.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UVM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11140</th>\n",
       "      <td>tcga-vd-a8kj</td>\n",
       "      <td>7289bc25-2551-4c6d-b2fa-c7e13859b9b3</td>\n",
       "      <td>tcga-vd-a8kj-10a</td>\n",
       "      <td>f470bbbf-57cb-4d19-9338-9e25ecfc6b96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19562.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UVM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11141 rows × 191 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      patient_barcode                               case_id    sample_barcode  \\\n",
       "0        tcga-or-a5ld  d885b720-e813-4828-a731-b3cdcbd20687  tcga-or-a5ld-10a   \n",
       "1        tcga-or-a5jd  240dca72-76e0-4d57-978d-1242e6e7ed68  tcga-or-a5jd-10b   \n",
       "2        tcga-or-a5jm  fb54458d-c373-46c2-841e-82663e13efaa  tcga-or-a5jm-10b   \n",
       "3        tcga-or-a5jz  94cdac79-2b4e-483a-8f17-3e0c9be589d0  tcga-or-a5jz-10a   \n",
       "4        tcga-ou-a5pi  a5be4e42-cb6c-4e67-8547-45dbd058bb95  tcga-ou-a5pi-10a   \n",
       "...               ...                                   ...               ...   \n",
       "11136    tcga-vd-aa8q  1d00217f-28df-4d91-aeb9-0d9a83a907e6  tcga-vd-aa8q-10a   \n",
       "11137    tcga-v4-a9ej  d9ccba9b-6bf0-44d9-8097-a97ab747f0bd  tcga-v4-a9ej-10a   \n",
       "11138    tcga-wc-a87u  4498044c-1607-403e-867e-befd7fad275a  tcga-wc-a87u-10a   \n",
       "11139    tcga-v4-a9eh  5e54122f-1c4d-45cc-8842-cd1c1df0eed8  tcga-v4-a9eh-10a   \n",
       "11140    tcga-vd-a8kj  7289bc25-2551-4c6d-b2fa-c7e13859b9b3  tcga-vd-a8kj-10a   \n",
       "\n",
       "                                sample_uuid  female  male   race  \\\n",
       "0      7f0378a8-e904-404b-91de-a70e73bda1b4     NaN   1.0  white   \n",
       "1      4fdc28a5-f64e-4d98-baac-befad27fb2a1     1.0   NaN  white   \n",
       "2      a3c80a41-bce1-4f72-8195-f1c699957c5f     1.0   NaN  white   \n",
       "3      a9c6d6e4-ab79-4a75-bd63-891647d82501     NaN   1.0  white   \n",
       "4      a37b1255-a17c-49e9-97f7-8525aee7438b     1.0   NaN  white   \n",
       "...                                     ...     ...   ...    ...   \n",
       "11136  d57d0601-a8eb-471e-9055-109caea0ecb9     NaN   1.0    NaN   \n",
       "11137  cef70e8e-da78-4eec-ab47-63a8d2be2cc4     NaN   1.0  white   \n",
       "11138  929e6183-8e31-44ee-a2d9-89b3e8861e01     NaN   1.0  white   \n",
       "11139  9256ff7e-5674-4e61-9fdc-947669e525c7     NaN   1.0  white   \n",
       "11140  f470bbbf-57cb-4d19-9338-9e25ecfc6b96     NaN   1.0    NaN   \n",
       "\n",
       "                    ethnicity  days_to_birth  alive  ...  primary_tumor  \\\n",
       "0      not hispanic or latino        19111.0    NaN  ...            1.0   \n",
       "1                         NaN        20989.0    3.0  ...            1.0   \n",
       "2                         NaN         9359.0    NaN  ...            1.0   \n",
       "3                         NaN        22218.0    2.0  ...            1.0   \n",
       "4      not hispanic or latino        19492.0    2.0  ...            1.0   \n",
       "...                       ...            ...    ...  ...            ...   \n",
       "11136                     NaN        26285.0    2.0  ...            1.0   \n",
       "11137  not hispanic or latino        13958.0    3.0  ...            1.0   \n",
       "11138  not hispanic or latino        26933.0    NaN  ...            1.0   \n",
       "11139      hispanic or latino        19573.0    2.0  ...            1.0   \n",
       "11140                     NaN        19562.0    3.0  ...            1.0   \n",
       "\n",
       "       blood_derived_normal  solid_tissue_normal  Proj_name  metastatic  \\\n",
       "0                       1.0                  NaN        ACC         NaN   \n",
       "1                       1.0                  NaN        ACC         NaN   \n",
       "2                       1.0                  NaN        ACC         NaN   \n",
       "3                       1.0                  NaN        ACC         NaN   \n",
       "4                       1.0                  NaN        ACC         NaN   \n",
       "...                     ...                  ...        ...         ...   \n",
       "11136                   1.0                  NaN        UVM         NaN   \n",
       "11137                   1.0                  NaN        UVM         NaN   \n",
       "11138                   1.0                  NaN        UVM         NaN   \n",
       "11139                   1.0                  NaN        UVM         NaN   \n",
       "11140                   1.0                  NaN        UVM         NaN   \n",
       "\n",
       "       recurrent_tumor  bone_marrow_normal  buccal_cell_normal  tumor  \\\n",
       "0                  NaN                 NaN                 NaN    NaN   \n",
       "1                  NaN                 NaN                 NaN    NaN   \n",
       "2                  NaN                 NaN                 NaN    NaN   \n",
       "3                  NaN                 NaN                 NaN    NaN   \n",
       "4                  NaN                 NaN                 NaN    NaN   \n",
       "...                ...                 ...                 ...    ...   \n",
       "11136              NaN                 NaN                 NaN    NaN   \n",
       "11137              NaN                 NaN                 NaN    NaN   \n",
       "11138              NaN                 NaN                 NaN    NaN   \n",
       "11139              NaN                 NaN                 NaN    NaN   \n",
       "11140              NaN                 NaN                 NaN    NaN   \n",
       "\n",
       "       additional_metastatic  \n",
       "0                        NaN  \n",
       "1                        NaN  \n",
       "2                        NaN  \n",
       "3                        NaN  \n",
       "4                        NaN  \n",
       "...                      ...  \n",
       "11136                    NaN  \n",
       "11137                    NaN  \n",
       "11138                    NaN  \n",
       "11139                    NaN  \n",
       "11140                    NaN  \n",
       "\n",
       "[11141 rows x 191 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "patient and cancer columns are case_id and Proj_name.\n",
      "Proj_name values are: ['ACC', 'BLCA', 'BRCA', 'CESC', 'CHOL', 'COAD', 'DLBC', 'ESCA', 'GBM', 'HNSC', 'KICH', 'KIRC', 'KIRP', 'LAML', 'LGG', 'LIHC', 'LUAD', 'LUSC', 'MESO', 'OV', 'PAAD', 'PCPG', 'PRAD', 'READ', 'SARC', 'SKCM', 'STAD', 'TGCT', 'THCA', 'THYM', 'UCEC', 'UCS', 'UVM']\n",
      "\n",
      "Patients per cancer:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cancer</th>\n",
       "      <th>Cancer_name</th>\n",
       "      <th>Num_patients</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>Breast invasive carcinoma</td>\n",
       "      <td>1095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GBM</td>\n",
       "      <td>Glioblastoma multiforme</td>\n",
       "      <td>594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OV</td>\n",
       "      <td>Ovarian serous cystadenocarcinoma</td>\n",
       "      <td>585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td>546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>KIRC</td>\n",
       "      <td>Kidney renal clear cell carcinoma</td>\n",
       "      <td>535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HNSC</td>\n",
       "      <td>Head and Neck squamous cell carcinoma</td>\n",
       "      <td>526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>LUAD</td>\n",
       "      <td>Lung adenocarcinoma</td>\n",
       "      <td>521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>LGG</td>\n",
       "      <td>Brain Lower Grade Glioma</td>\n",
       "      <td>513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>THCA</td>\n",
       "      <td>Thyroid carcinoma</td>\n",
       "      <td>505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LUSC</td>\n",
       "      <td>Lung squamous cell carcinoma</td>\n",
       "      <td>502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PRAD</td>\n",
       "      <td>Prostate adenocarcinoma</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SKCM</td>\n",
       "      <td>Skin Cutaneous Melanoma</td>\n",
       "      <td>470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>COAD</td>\n",
       "      <td>Colon adenocarcinoma</td>\n",
       "      <td>459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>STAD</td>\n",
       "      <td>Stomach adenocarcinoma</td>\n",
       "      <td>443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>Bladder Urothelial Carcinoma</td>\n",
       "      <td>412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>LIHC</td>\n",
       "      <td>Liver hepatocellular carcinoma</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CESC</td>\n",
       "      <td>Cervical squamous cell carcinoma and endocervi...</td>\n",
       "      <td>307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>KIRP</td>\n",
       "      <td>Kidney renal papillary cell carcinoma</td>\n",
       "      <td>291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SARC</td>\n",
       "      <td>Sarcoma</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>LAML</td>\n",
       "      <td>Acute Myeloid Leukemia</td>\n",
       "      <td>200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ESCA</td>\n",
       "      <td>Esophageal carcinoma</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>PAAD</td>\n",
       "      <td>Pancreatic adenocarcinoma</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PCPG</td>\n",
       "      <td>Pheochromocytoma and Paraganglioma</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>READ</td>\n",
       "      <td>Rectum adenocarcinoma</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TGCT</td>\n",
       "      <td>Testicular Germ Cell Tumors</td>\n",
       "      <td>134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>THYM</td>\n",
       "      <td>Thymoma</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>KICH</td>\n",
       "      <td>Kidney Chromophobe</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ACC</td>\n",
       "      <td>Adrenocortical carcinoma</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MESO</td>\n",
       "      <td>Mesothelioma</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>UVM</td>\n",
       "      <td>Uveal Melanoma</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>UCS</td>\n",
       "      <td>Uterine Carcinosarcoma</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>DLBC</td>\n",
       "      <td>Lymphoid Neoplasm Diffuse Large B-cell Lymphoma</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>CHOL</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Cancer                                        Cancer_name  Num_patients\n",
       "1    BRCA                          Breast invasive carcinoma          1095\n",
       "2     GBM                            Glioblastoma multiforme           594\n",
       "3      OV                  Ovarian serous cystadenocarcinoma           585\n",
       "4    UCEC               Uterine Corpus Endometrial Carcinoma           546\n",
       "5    KIRC                  Kidney renal clear cell carcinoma           535\n",
       "6    HNSC              Head and Neck squamous cell carcinoma           526\n",
       "7    LUAD                                Lung adenocarcinoma           521\n",
       "8     LGG                           Brain Lower Grade Glioma           513\n",
       "9    THCA                                  Thyroid carcinoma           505\n",
       "10   LUSC                       Lung squamous cell carcinoma           502\n",
       "11   PRAD                            Prostate adenocarcinoma           500\n",
       "12   SKCM                            Skin Cutaneous Melanoma           470\n",
       "13   COAD                               Colon adenocarcinoma           459\n",
       "14   STAD                             Stomach adenocarcinoma           443\n",
       "15   BLCA                       Bladder Urothelial Carcinoma           412\n",
       "16   LIHC                     Liver hepatocellular carcinoma           377\n",
       "17   CESC  Cervical squamous cell carcinoma and endocervi...           307\n",
       "18   KIRP              Kidney renal papillary cell carcinoma           291\n",
       "19   SARC                                            Sarcoma           261\n",
       "20   LAML                             Acute Myeloid Leukemia           200\n",
       "21   ESCA                               Esophageal carcinoma           185\n",
       "22   PAAD                          Pancreatic adenocarcinoma           185\n",
       "23   PCPG                 Pheochromocytoma and Paraganglioma           179\n",
       "24   READ                              Rectum adenocarcinoma           170\n",
       "25   TGCT                        Testicular Germ Cell Tumors           134\n",
       "26   THYM                                            Thymoma           124\n",
       "27   KICH                                 Kidney Chromophobe           113\n",
       "28    ACC                           Adrenocortical carcinoma            92\n",
       "29   MESO                                       Mesothelioma            87\n",
       "30    UVM                                     Uveal Melanoma            80\n",
       "31    UCS                             Uterine Carcinosarcoma            57\n",
       "32   DLBC    Lymphoid Neoplasm Diffuse Large B-cell Lymphoma            48\n",
       "33   CHOL                                 Cholangiocarcinoma            45"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Patients per therapy:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Therapy</th>\n",
       "      <th>Num_patients</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALL_patients</td>\n",
       "      <td>11086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NO_therapy</td>\n",
       "      <td>6848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALL_therapies</td>\n",
       "      <td>4238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chemotherapy</td>\n",
       "      <td>3748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>carboplatin</td>\n",
       "      <td>962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>carmustine</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>aromasin</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>temsirolimus</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>bcnu</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>carboplatinum</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Therapy  Num_patients\n",
       "1    ALL_patients         11086\n",
       "2      NO_therapy          6848\n",
       "3   ALL_therapies          4238\n",
       "4    chemotherapy          3748\n",
       "5     carboplatin           962\n",
       "..            ...           ...\n",
       "74     carmustine            21\n",
       "75       aromasin            20\n",
       "76   temsirolimus            20\n",
       "77           bcnu            20\n",
       "78  carboplatinum            20\n",
       "\n",
       "[78 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"\"\" \n",
    "Debug:Show\n",
    "\"\"\"\n",
    "print(f\"Clinical data\")\n",
    "display(Dsa_data.get_clinical_data())\n",
    "print(f\"patient and cancer columns are {Dsa_data.patient_ID_col} and {Dsa_data.cancer_col}.\")\n",
    "print(f\"{Dsa_data.cancer_col} values are: {Dsa_data.get_clinical_data()[Dsa_data.cancer_col].unique().tolist()}\")\n",
    "\n",
    "print(f\"\\nPatients per cancer:\")\n",
    "a = [(cnr, df_c.shape[0]) for cnr, df_c in Dsa_data.get_clinical_data().groupby(by=Dsa_data.cancer_col)]\n",
    "df_cancer_prevalence = pd.DataFrame({'Cancer': [x[0] for x in a], 'Num_patients': [x[1] for x in a]}).sort_values('Num_patients', ascending=False).reset_index(drop=True)\n",
    "df_cancer_prevalence.index += 1\n",
    "df_cancer_prevalence['Cancer_name'] = df_cancer_prevalence.apply(lambda x: gdcu.Cancer_abbrv2name[f\"TCGA_{x['Cancer']}\"], axis=1)\n",
    "df_cancer_prevalence = df_cancer_prevalence[['Cancer', 'Cancer_name', 'Num_patients']]\n",
    "display(df_cancer_prevalence)\n",
    "\n",
    "print(\"\\nPatients per therapy:\")\n",
    "df_clinical_therapies_dict = Dsa_data.df_clinical_therapies_dict\n",
    "df_therapy_prevalence = pd.DataFrame({'Therapy': list(df_clinical_therapies_dict.keys()), 'Num_patients': [x.shape[0] for x in df_clinical_therapies_dict.values()]}).sort_values('Num_patients', ascending=False).reset_index(drop=True)\n",
    "df_therapy_prevalence.index += 1\n",
    "display(df_therapy_prevalence)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parameters\n",
    "Set the testing parameters in this section.\n",
    "\n",
    "Two analysis modes are supported:\n",
    "- Analysing a given list of arbitrary mutations, by setting the `muts_ids` list.\n",
    "- Analysing all (possibly filtered) mutations of a given gene, by setting the `gene` string.\n",
    "\n",
    "**`muts_ids` takes precedence over `gene`, i.e. if `muts_ids` is not an empty list, these mutations will be analyzd, else (`muts_ids` is an empty list), `gene` mutations will be analyzed.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" \n",
    "Input\n",
    "\"\"\"\n",
    "# List of mutations\n",
    "#====================\n",
    "# list of mutation IDs (with mut_id_cols separated by mut_id_sep)\n",
    "# muts_ids: list[str] = [\n",
    "#     'TPX2:chr20:Splice_Region:31770476:G:T',\n",
    "#     #'FAM193A:chr4:Splice_Site:2695131:T:G',\n",
    "#     #'MET:chr7:Splice_Site:116739948:A:T',\n",
    "#     #'PIK3CA:chr3:Intron:179203524:G:C',\n",
    "#     #'FLT3:chr13:splice_donor_variant:28034299:A:T',\n",
    "#     'CDON:chr11:Splice_Site:126021246:A:C'\n",
    "# ]\n",
    "\n",
    "# update set of splice-altering mutations (with prevalence>=10% in certain cancer types)\n",
    "muts_ids: list[str] = [\n",
    "    'TPX2:chr20:Splice_Region:31770476:G:T',\n",
    "    'MTMR11:chr1:Splice_Site:149935297:A:C',\n",
    "    'GSN:chr9:Splice_Site:121300007:T:G',\n",
    "    'SPEG:chr2:Splice_Site:219477787:T:G',\n",
    "    'HERC3:chr4:Splice_Site:88655306:T:G',\n",
    "    'CHL1:chr3:Splice_Site:401624:A:T',\n",
    "    \n",
    "    'TRPC4AP:chr20:Splice_Region:35054969:T:G',\n",
    "    'APBA2:chr15:Splice_Site:29076093:T:G',\n",
    "    'NBPF3:chr1:Splice_Site:21473586:G:T',\n",
    "    'RAD51D:chr17:Splice_Site:35118499:A:C',\n",
    "]\n",
    "\n",
    "# mutations affecting mRNA levels (see Explore_GDC/Reports/GDC_expr_effect_sgnf_CDx_genes_scores_gap-1.xlsx)\n",
    "# muts_ids: list[str] = [\n",
    "#     'EGFR:chr7:Intron:55201370:TC:-',\n",
    "#     'BRCA1:chr17:Intron:43090922:CA:-',\n",
    "#     'IDH1:chr2:Intron:208245452:A:-',\n",
    "#     'PIK3CA:chr3:Intron:179203524:G:C',  # note that this mutation is also listed above\n",
    "#     'MET:chr7:Intron:116769622:TTT:-',\n",
    "#     'ROS1:chr6:Intron:117310301:TAA:-',\n",
    "#     'PIK3CA:chr3:Intron:179224066:T:-',\n",
    "#     'PIK3CA:chr3:Intron:179220948:TA:-'\n",
    "# ]\n",
    "\n",
    "# mutations affecting methylation\n",
    "# muts_ids: list[str] = [\n",
    "#     \"TPM4:chr19:5'Flank:16076004:C:-\",\n",
    "#     \"XPC:chr3:5'UTR:14178635:C:A\",\n",
    "#     \"THRA:chr17:Intron:40089367:G:C\",\n",
    "#     \"ID1:chr20:5'Flank:31605268:A:C\",\n",
    "#     \"TCP1:chr6:5'Flank:159790389:A:G\",\n",
    "#     \"BCL9L:chr11:Intron:118903196:AG:-\",\n",
    "#     \"TCP1:chr6:5'Flank:159790413:C:T\",\n",
    "# ]\n",
    "\n",
    "# test individual mutation\n",
    "# muts_ids: list[str] = [\n",
    "#     #'PIK3CA:chr3:Intron:179203524:G:C',\n",
    "#     #'TPX2:chr20:Splice_Region:31770476:G:T',\n",
    "#     #'APBA2:chr15:Splice_Site:29076093:T:G',\n",
    "#     #'GLG1:chr16:Splice_Site:74483125:C:T',\n",
    "#     #'MYOF:chr10:Splice_Site:93379861:A:C',\n",
    "#     #'MTMR11:chr1:Splice_Site:149935297:A:C',\n",
    "#     #'HERC3:chr4:Splice_Site:88655306:T:G',\n",
    "#     #'CAD:chr2:Splice_Site:27218017:G:T',\n",
    "#     #'ATP1A2:chr1:Splice_Site:160135297:T:G',\n",
    "#     #'DYRK4:chr12:Splice_Site:4591300:T:G',\n",
    "#     #'MSH5:chr6:Splice_Site:31759977:T:G',\n",
    "#     #'EMC1:chr1:Splice_Site:19243612:A:C',\n",
    "#     #'RAD51D:chr17:Splice_Site:35118499:A:C',\n",
    "#     #'CDON:chr11:Splice_Site:126021246:A:C',\n",
    "\n",
    "#     #'EGFR:chr7:Intron:55201370:TC:-',\n",
    "#     #'POLR3B:chr12:Splice_Site:106369685:T:G',\n",
    "#     #'SPEG:chr2:Splice_Site:219477787:T:G',\n",
    "#     #'ATP1A2:chr1:Splice_Site:160135297:T:G',\n",
    "#     #'PRRC2A:chr6:Splice_Site:31629336:T:G',\n",
    "#     #'AASS:chr7:Splice_Site:122126373:A:C',\n",
    "#     #'RAD51D:chr17:Splice_Site:35118499:A:C',\n",
    "#     #'TSHR:chr14:Intron:81096719:G:T',\n",
    "#     #'SPTBN1:chr2:Splice_Site:54599245:T:G',\n",
    "#     \"BCL9L:chr11:Intron:118903196:AG:-\",\n",
    "# ]\n",
    "\n",
    "#muts_file = pathlib.Path(si.ROOT_path) / 'OncoD' / 'Experiments' / 'Splicing_2023' / 'candidates' / 'splice_aug_2023' / 'splice_list_10_15.xlsx'\n",
    "#muts_file = pathlib.Path(si.ROOT_path) / 'OncoD' / 'Experiments' / 'Splicing_2023' / 'candidates' / 'splice_jan_2024' / 'splice_list_1_22.xlsx'\n",
    "#muts_ids: list[str] = pd.read_excel(muts_file)['mut_id'].unique().tolist()\n",
    "\n",
    "# Or, mutations of a gene. (set muts_ids = [] for this to take effect)\n",
    "# ====================================================================\n",
    "#muts_ids: list[str] = []; gene: str = 'MET' #'SPEG' # 'TPX2' # 'TPX2'\n",
    "filter_data: dict = {\n",
    "    'mut_id_col_to_add': 'mut_id',  # name of a column to add to the mutation data holding the mutation ID\n",
    "    'filter_silent': True,  # choose only gene's silent mutations\n",
    "    'filter_TCGA_projects': True,  # choose only gene's TCGA projects (should be set to True)\n",
    "    'mutation_min_affected_number_of_patients': 10  # choose only mutations affecting at least this number of patients\n",
    "}\n",
    "\n",
    "\"\"\" \n",
    "Test configuration\n",
    "\"\"\"\n",
    "Selected_test_config_name: str = 'MCT-mCT'  # patients with mutation vs patients without mutation, over each cancer and therapy type\n",
    "#Selected_test_config_name: str = 'MCT-mCT,MCt-mCt'  # 2 tests.\n",
    "#Selected_test_config_name: str = 'MCT-MCt,mCT-mCt' # 2 tests: first patients with mutation and with therapy vs patients with mutation and without therapy, over each cancer and therapy type,\n",
    "                                                   # second test: patients without mutation and with therapy vs patients without mutation and without therapy, over each cancer and therapy type\n",
    "#Selected_test_config_name: str = 'McT-mcT'  # patients with mutation vs patients without mutation, pan-cancer over each therapy type\n",
    "\n",
    "Main_test_index: int = 1  # see above for description\n",
    "\n",
    "# minimum number of patients in each group required to evaluate survival\n",
    "Min_group_size: int = 6 # 10 # 8\n",
    "\n",
    "\"\"\"\n",
    "Output file name\n",
    "\"\"\"\n",
    "save_result_base_name: pathlib.Path = results_path / f\"DSA_sgnf_methylation_{ut.DS_test_group_cnfg.test_name2indication_name(Selected_test_config_name).replace(',', '_')}_splicemuts_B3\"\n",
    "\n",
    "#save_result_base_name: pathlib.Path = results_path / f\"DSA_sgnf_methylation_{ut.DS_test_group_cnfg.test_name2indication_name(Selected_test_config_name).replace(',', '_')}_splicemuts\"\n",
    "\n",
    "#save_result_base_name: pathlib.Path = results_path / \"DSA_sgnf_methylation_dummy_vv\"\n",
    "#save_result_base_name: pathlib.Path = results_path / f\"DSA_sgnf_methylation_{ut.DS_test_group_cnfg.test_name2indication_name(Selected_test_config_name).replace(',', '_')}_v4\"  # code will add .pkl and .xlsx\n",
    "\n",
    "#save_result_base_name: pathlib.Path = results_path / f\"DSA_sgnf_individual_{Taut.DS_test_group_cnfg.test_name2indication_name(Selected_test_config_name).replace(',', '_')}_v4\"  # code will add .pkl and .xlsx\n",
    "\n",
    "#save_result_base_name: pathlib.Path = results_path / f\"DSA_sgnf_mrnalevels_{Taut.DS_test_group_cnfg.test_name2indication_name(Selected_test_config_name).replace(',', '_')}_v4\"  # code will add .pkl and .xlsx\n",
    "#save_result_base_name: pathlib.Path = results_path / f\"DSA_sgnf_5muts_{Taut.DS_test_group_cnfg.test_name2indication_name(Selected_test_config_name).replace(',', '_')}_v4\"  # code will add .pkl and .xlsx\n",
    "#save_result_base_name = results_path / f\"DSA_sgnf_{muts_file.stem}_{Taut.DS_test_group_cnfg.test_name2indication_name(Selected_test_config_name).replace(',', '_')}_v4\"  # code will add .pkl and .xlsx\n",
    "#save_result_base_name: pathlib.Path = results_path / f\"DSA_sgnf_{gene}_{Taut.DS_test_group_cnfg.test_name2indication_name(Selected_test_config_name).replace(',', '_')}_v4\"  # code will add .pkl and .xlsx\n",
    "\n",
    "\"\"\" \n",
    "Debug level\n",
    "\"\"\"\n",
    "# control prints in the Analyze section\n",
    "debug_level: int = 1  # 0 (no prints), 1 (significant prints), 2 (more prints), 3 (all prints)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" \n",
    "Filtering few cancers and therapies.\n",
    "\n",
    "Use this cell to select a subset of cancer types and therapies to analyze.\n",
    "\"\"\"\n",
    "Filtering: bool = False\n",
    "\n",
    "#Filtering_use_cancers: list = ['OV']  # cancer types to analyze (without the 'TCGA_'. E.g., 'OV')\n",
    "#Filtering_use_therapies: list = ['paclitaxel'] + [Dsa_data.None_therapy_name, Dsa_data.All_name]  # therapies to analyze (keys in df_clinical_therapies_dict)\n",
    "# Filtering_use_therapies: list = df_therapy_prevalence['Therapy'].tolist()  # all therapies/drugs\n",
    "\n",
    "# analyzing only the all patients group (i.e., not analysing the different therapies)\n",
    "Filtering_use_cancers: list = df_cancer_prevalence['Cancer'].tolist()\n",
    "Filtering_use_therapies: list = [Dsa_data.None_therapy_name, Dsa_data.All_name, Dsa_data.All_therapies_name]\n",
    "# ============================================================================================================================================\n",
    "if Filtering:\n",
    "    print(f\"Performing filtering:\\n=====================\\n{Filtering_use_cancers=}\\n{Filtering_use_therapies=}\")\n",
    "\n",
    "    # filtering cancer types\n",
    "    Ignore_cancers = {f\"TCGA_{x}\" for x in df_cancer_prevalence['Cancer'].unique().tolist() if x not in Filtering_use_cancers}\n",
    "\n",
    "    # filtering therapies\n",
    "    tmp_df_clinical_therapies_dict: dict = {k: v for k, v in Dsa_data.df_clinical_therapies_dict.items() if k in Filtering_use_therapies}\n",
    "    org_df_clinical_therapies_dict = Dsa_data.df_clinical_therapies_dict\n",
    "    Dsa_data.df_clinical_therapies_dict = tmp_df_clinical_therapies_dict\n",
    "\n",
    "    print(f\"New:\\n{Ignore_cancers=}\")\n",
    "    for k, v in Dsa_data.df_clinical_therapies_dict.items():\n",
    "        print(f\"{k}: {v.shape[0]} cases\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Updating p value threshold - e.g., when we want to log also large p values (i.e. insignificant results).\n",
    "# Dsa_compute.P_val_threshold = 0.99\n",
    "# Dsa_compute.P_val_threshold = 0.06\n",
    "# print(f\"UPDATED {Dsa_compute.P_val_threshold=} !!!!!!!!!!!!!!!!!!!!!!!!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test_type='arbitrary_mutations', Selected_test_config_name='MCT-mCT' (tests_config=[[(True, True, True), (False, True, True)]]), Main_test_index=1, Min_group_size=6,\n",
      "Dsa_compute.Enable_sampling_analysis=True, Dsa_compute.Sampling_p_val_method='logrank', Dsa_compute.Sampling_average_method='median'\n",
      "Ignore_therapies={'NO_therapy', 'ALL_therapies'}\n",
      "debug_level=1\n",
      "save_result_base_name=PosixPath('/Users/yoramzarai/work/mystuff/Ramot/Projects/Cancer_mut/Explore_kaplan_meier/Develop1/Reports/DSA_results/DSA_sgnf_methylation_111-011_splicemuts_B3').\n"
     ]
    }
   ],
   "source": [
    "\"\"\" \n",
    "Show configuration\n",
    "\"\"\"\n",
    "Test_type: str = 'arbitrary_mutations' if muts_ids else 'gene_mutations'\n",
    "\n",
    "# logrank parameters \n",
    "Logrank_kwargs: dict = {'weightings': Dsa_compute.Logrank_weightings if Dsa_compute.Logrank_weightings else None, 'p': Dsa_compute.Logrank_p, 'q': Dsa_compute.Logrank_q}\n",
    "\n",
    "# get tests configuration\n",
    "try:\n",
    "    tests_config = ut.DS_test_group_cnfg.DS_name2config(Selected_test_config_name)  # ut.DS_test_name2config[Selected_test_config_name]\n",
    "except KeyError:\n",
    "    print(f\"{Selected_test_config_name=} is not supported !! Please revise 'Selected_test_config_name' (see DS_test_name2config in utils/TCGA_survival_utils.py).\")\n",
    "    raise\n",
    "\n",
    "print(f\"{Test_type=}, {Selected_test_config_name=} ({tests_config=}), {Main_test_index=}, {Min_group_size=},\\n{Dsa_compute.Enable_sampling_analysis=}, {Dsa_compute.Sampling_p_val_method=}, {Dsa_compute.Sampling_average_method=}\\n{Ignore_therapies=}\\n{debug_level=}\\n{save_result_base_name=}.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analyzing 10 mutations from 10 genes:\n",
      "\n",
      "1. Processing gene='TPX2' ...\n",
      "*****[MCT-mCT, TPX2:chr20:Splice_Region:31770476:G:T, TCGA_LAML, ALL_patients | G1_size=21, G2_size=167, pv_cox=0.001933, pv_logrank=0.001310, AUC1=1.26, AUC2=2.62 | pv_samp=0.072003, AUC_samp=2.30]*****\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.181. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*****[MCT-mCT, TPX2:chr20:Splice_Region:31770476:G:T, TCGA_LUAD, ALL_patients | G1_size=7, G2_size=505, pv_cox=0.033984, pv_logrank=0.027587, AUC1=2.75, AUC2=7.52 | pv_samp=0.261976, AUC_samp=3.57]*****\n",
      "*****[MCT-mCT, TPX2:chr20:Splice_Region:31770476:G:T, TCGA_OV, taxol | G1_size=10, G2_size=243, pv_cox=0.017172, pv_logrank=0.013694, AUC1=2.47, AUC2=4.25 | pv_samp=0.103567, AUC_samp=3.96]*****\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.141. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.133. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2. Processing gene='MTMR11' ...\n",
      "*****[MCT-mCT, MTMR11:chr1:Splice_Site:149935297:A:C, TCGA_LAML, ALL_patients | G1_size=17, G2_size=171, pv_cox=0.001204, pv_logrank=0.000788, AUC1=0.97, AUC2=2.81 | pv_samp=0.077004, AUC_samp=2.35]*****\n",
      "*****[MCT-mCT, MTMR11:chr1:Splice_Site:149935297:A:C, TCGA_BRCA, doxorubicin | G1_size=12, G2_size=169, pv_cox=0.009389, pv_logrank=0.003407, AUC1=5.77, AUC2=10.55 | pv_samp=0.113555, AUC_samp=8.70]*****\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.267. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*****[MCT-mCT, MTMR11:chr1:Splice_Site:149935297:A:C, TCGA_BRCA, paclitaxel | G1_size=6, G2_size=127, pv_cox=0.023595, pv_logrank=0.010424, AUC1=5.40, AUC2=9.99 | pv_samp=0.197761, AUC_samp=8.25]*****\n",
      "*****[MCT-mCT, MTMR11:chr1:Splice_Site:149935297:A:C, TCGA_BRCA, arimidex | G1_size=7, G2_size=154, pv_cox=0.008291, pv_logrank=0.001649, AUC1=4.47, AUC2=10.05 | pv_samp=0.233043, AUC_samp=6.20]*****\n",
      "*****[MCT-mCT, MTMR11:chr1:Splice_Site:149935297:A:C, TCGA_OV, taxol | G1_size=10, G2_size=243, pv_cox=0.015829, pv_logrank=0.012441, AUC1=2.26, AUC2=4.26 | pv_samp=0.108341, AUC_samp=3.97]*****\n",
      "*****[MCT-mCT, MTMR11:chr1:Splice_Site:149935297:A:C, TCGA_OV, carboplatin | G1_size=28, G2_size=433, pv_cox=0.018204, pv_logrank=0.016584, AUC1=3.53, AUC2=5.05 | pv_samp=0.134091, AUC_samp=4.75]*****\n",
      "*****[MCT-mCT, MTMR11:chr1:Splice_Site:149935297:A:C, TCGA_OV, chemotherapy | G1_size=32, G2_size=495, pv_cox=0.040670, pv_logrank=0.038807, AUC1=3.75, AUC2=5.06 | pv_samp=0.163818, AUC_samp=4.82]*****\n",
      "3. Processing gene='GSN' ...\n",
      "*****[MCT-mCT, GSN:chr9:Splice_Site:121300007:T:G, TCGA_PAAD, ALL_patients | G1_size=7, G2_size=178, pv_cox=0.021759, pv_logrank=0.016230, AUC1=0.87, AUC2=2.85 | pv_samp=0.170529, AUC_samp=1.96]*****\n",
      "*****[MCT-mCT, GSN:chr9:Splice_Site:121300007:T:G, TCGA_LIHC, ALL_patients | G1_size=53, G2_size=323, pv_cox=0.004816, pv_logrank=0.003673, AUC1=7.68, AUC2=4.58 | pv_samp=0.021120, AUC_samp=4.38]*****\n",
      "4. Processing gene='SPEG' ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.259. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.279. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.237. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.244. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*****[MCT-mCT, SPEG:chr2:Splice_Site:219477787:T:G, TCGA_PAAD, ALL_patients | G1_size=28, G2_size=157, pv_cox=0.033273, pv_logrank=0.031423, AUC1=1.59, AUC2=3.04 | pv_samp=0.106920, AUC_samp=2.77]*****\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.191. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5. Processing gene='HERC3' ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.325. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.291. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.193. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.153. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.152. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.181. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*****[MCT-mCT, HERC3:chr4:Splice_Site:88655306:T:G, TCGA_CESC, ALL_patients | G1_size=9, G2_size=298, pv_cox=0.026654, pv_logrank=0.020506, AUC1=2.77, AUC2=10.41 | pv_samp=0.190505, AUC_samp=6.32]*****\n",
      "*****[MCT-mCT, HERC3:chr4:Splice_Site:88655306:T:G, TCGA_OV, doxil | G1_size=10, G2_size=111, pv_cox=0.000956, pv_logrank=0.000479, AUC1=2.63, AUC2=4.31 | pv_samp=0.023720, AUC_samp=4.16]*****\n",
      "*****[MCT-mCT, HERC3:chr4:Splice_Site:88655306:T:G, TCGA_GBM, ALL_patients | G1_size=61, G2_size=533, pv_cox=0.035436, pv_logrank=0.035115, AUC1=1.28, AUC2=1.73 | pv_samp=0.217519, AUC_samp=1.71]*****\n",
      "*****[MCT-mCT, HERC3:chr4:Splice_Site:88655306:T:G, TCGA_GBM, chemotherapy | G1_size=44, G2_size=377, pv_cox=0.032030, pv_logrank=0.031257, AUC1=1.46, AUC2=2.06 | pv_samp=0.172195, AUC_samp=2.02]*****\n",
      "*****[MCT-mCT, HERC3:chr4:Splice_Site:88655306:T:G, TCGA_GBM, temozolomide | G1_size=24, G2_size=145, pv_cox=0.050798, pv_logrank=0.048572, AUC1=1.54, AUC2=2.35 | pv_samp=0.160323, AUC_samp=2.32]*****\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.261. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6. Processing gene='CHL1' ...\n",
      "7. Processing gene='TRPC4AP' ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.381. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.337. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*****[MCT-mCT, TRPC4AP:chr20:Splice_Region:35054969:T:G, TCGA_BRCA, taxol | G1_size=7, G2_size=82, pv_cox=0.068597, pv_logrank=0.045246, AUC1=5.99, AUC2=10.54 | pv_samp=0.208668, AUC_samp=8.03]*****\n",
      "*****[MCT-mCT, TRPC4AP:chr20:Splice_Region:35054969:T:G, TCGA_LAML, ALL_patients | G1_size=12, G2_size=176, pv_cox=0.023575, pv_logrank=0.021208, AUC1=1.07, AUC2=2.75 | pv_samp=0.271521, AUC_samp=2.15]*****\n",
      "*****[MCT-mCT, TRPC4AP:chr20:Splice_Region:35054969:T:G, TCGA_KIRC, ALL_patients | G1_size=26, G2_size=509, pv_cox=0.010285, pv_logrank=0.004951, AUC1=11.01, AUC2=6.92 | pv_samp=0.024113, AUC_samp=6.34]*****\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.164. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.163. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*****[MCT-mCT, TRPC4AP:chr20:Splice_Region:35054969:T:G, TCGA_GBM, ALL_patients | G1_size=30, G2_size=564, pv_cox=0.034632, pv_logrank=0.033391, AUC1=1.10, AUC2=1.71 | pv_samp=0.265122, AUC_samp=1.61]*****\n",
      "8. Processing gene='APBA2' ...\n",
      "*****[MCT-mCT, APBA2:chr15:Splice_Site:29076093:T:G, TCGA_BRCA, anastrozole | G1_size=9, G2_size=62, pv_cox=0.074982, pv_logrank=0.047445, AUC1=7.36, AUC2=12.49 | pv_samp=0.260492, AUC_samp=9.08]*****\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.370. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.292. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.375. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.288. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.349. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*****[MCT-mCT, APBA2:chr15:Splice_Site:29076093:T:G, TCGA_CESC, ALL_patients | G1_size=12, G2_size=295, pv_cox=0.011638, pv_logrank=0.008117, AUC1=4.21, AUC2=10.32 | pv_samp=0.118679, AUC_samp=6.55]*****\n",
      "*****[MCT-mCT, APBA2:chr15:Splice_Site:29076093:T:G, TCGA_OV, doxil | G1_size=7, G2_size=114, pv_cox=0.021582, pv_logrank=0.016619, AUC1=2.71, AUC2=4.24 | pv_samp=0.101383, AUC_samp=4.05]*****\n",
      "*****[MCT-mCT, APBA2:chr15:Splice_Site:29076093:T:G, TCGA_OV, taxol | G1_size=13, G2_size=240, pv_cox=0.001315, pv_logrank=0.000751, AUC1=2.31, AUC2=4.29 | pv_samp=0.022474, AUC_samp=3.95]*****\n",
      "*****[MCT-mCT, APBA2:chr15:Splice_Site:29076093:T:G, TCGA_OV, carboplatin | G1_size=36, G2_size=425, pv_cox=0.047845, pv_logrank=0.046616, AUC1=3.71, AUC2=5.05 | pv_samp=0.185163, AUC_samp=4.87]*****\n",
      "*****[MCT-mCT, APBA2:chr15:Splice_Site:29076093:T:G, TCGA_OV, chemotherapy | G1_size=44, G2_size=483, pv_cox=0.027626, pv_logrank=0.026637, AUC1=3.70, AUC2=5.10 | pv_samp=0.153161, AUC_samp=4.88]*****\n",
      "*****[MCT-mCT, APBA2:chr15:Splice_Site:29076093:T:G, TCGA_LUSC, ALL_patients | G1_size=37, G2_size=459, pv_cox=0.000910, pv_logrank=0.000702, AUC1=2.78, AUC2=6.11 | pv_samp=0.029227, AUC_samp=5.62]*****\n",
      "*****[MCT-mCT, APBA2:chr15:Splice_Site:29076093:T:G, TCGA_LUSC, cisplatin | G1_size=6, G2_size=65, pv_cox=0.032537, pv_logrank=0.023375, AUC1=1.79, AUC2=5.80 | pv_samp=0.215865, AUC_samp=3.99]*****\n",
      "9. Processing gene='NBPF3' ...\n",
      "*****[MCT-mCT, NBPF3:chr1:Splice_Site:21473586:G:T, TCGA_STAD, ALL_patients | G1_size=13, G2_size=425, pv_cox=0.003426, pv_logrank=0.001867, AUC1=0.92, AUC2=4.62 | pv_samp=0.028478, AUC_samp=2.78]*****\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.296. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.141. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10. Processing gene='RAD51D' ...\n",
      "*****[MCT-mCT, RAD51D:chr17:Splice_Site:35118499:A:C, TCGA_LAML, ALL_patients | G1_size=8, G2_size=180, pv_cox=0.000612, pv_logrank=0.000226, AUC1=0.83, AUC2=2.74 | pv_samp=0.090836, AUC_samp=2.01]*****\n",
      "*****[MCT-mCT, RAD51D:chr17:Splice_Site:35118499:A:C, TCGA_OV, paclitaxel | G1_size=7, G2_size=189, pv_cox=0.022906, pv_logrank=0.018025, AUC1=3.35, AUC2=6.09 | pv_samp=0.173678, AUC_samp=5.30]*****\n",
      "*****[MCT-mCT, RAD51D:chr17:Splice_Site:35118499:A:C, TCGA_GBM, temodar | G1_size=20, G2_size=143, pv_cox=0.018355, pv_logrank=0.016821, AUC1=1.16, AUC2=1.86 | pv_samp=0.059532, AUC_samp=1.69]*****\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/utils/__init__.py:1122: ConvergenceWarning: Column group1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['E'].astype(bool)\n",
      ">>> print(df.loc[events, 'group1'].var())\n",
      ">>> print(df.loc[~events, 'group1'].var())\n",
      "\n",
      "A very low variance means that the column group1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/yoramzarai/work/Ramot/Projects/Cancer_mut/venv311/lib/python3.11/site-packages/lifelines/fitters/coxph_fitter.py:1604: ConvergenceWarning: Newton-Rhaphson convergence completed successfully but norm(delta) is still high, 0.174. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "\"\"\" \n",
    "Run\n",
    "\"\"\"\n",
    "match Test_type:\n",
    "    case 'arbitrary_mutations':\n",
    "        # rearrange all mutations of the same gene in gene_muts_dict[gene]\n",
    "        gene_muts_dict = ut.create_gene_mutations_dict(muts_ids, Dsa_Mut)\n",
    "        print(f\"Analyzing {len(muts_ids):,} mutations from {len(gene_muts_dict):,} genes:\\n\")\n",
    "\n",
    "        gene_results = ut.process_muts_survival_tests_wrapper(tests_config, gene_muts_dict,\n",
    "                                                              Dsa_Mut, Dsa_data, Dsa_compute,\n",
    "                                                              Ignore_therapies, Ignore_cancers,\n",
    "                                                              Min_group_size, Logrank_kwargs, \n",
    "                                                              debug_level=debug_level, main_test_index=Main_test_index)\n",
    "        print(\"\\nDone.\")\n",
    "    case 'gene_mutations':\n",
    "        print(f\"Analyzing {gene}'s mutations using {filter_data=}:\\n\")\n",
    "\n",
    "        results, _, _ = ut.process_gene_all_muts_survival_tests_wrapper(\n",
    "            tests_config, gene, filter_data,\n",
    "            Dsa_Mut, Dsa_data, Dsa_compute,\n",
    "            Ignore_therapies, Ignore_cancers,\n",
    "            Min_group_size, Logrank_kwargs,\n",
    "            debug_level=debug_level, main_test_index=Main_test_index)\n",
    "        \n",
    "        gene_results = {}\n",
    "        if results:\n",
    "            print(f\"Found significant results in {gene=}\")\n",
    "            gene_results[gene] = results\n",
    "        print(\"\\nDone.\")\n",
    "    case _:\n",
    "        pass\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# View Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" \n",
    "Show informative results (as dataframe)\n",
    "\"\"\"\n",
    "if not (df_sgnf_results := ut.create_survival_results_dataframe(gene_results)).empty:\n",
    "    display(df_sgnf_results)\n",
    "    #display(df_sgnf_results.iloc[279-6:279-6+30].reset_index(drop=True))\n",
    "else:\n",
    "    print(\"No significant results.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" \n",
    "Use this to look for significant results of a X-X test in cases of multiple DS tests.\n",
    "\n",
    "Specifically, we are looking for a significant X test (i.e. a test with X in both groups) that does NOT have a matching test\n",
    "(i.e. a test where at least one of the groups has no X). \n",
    "\"\"\"\n",
    "save_interesting_results: bool = False\n",
    "# ====================================\n",
    "\n",
    "if (len(tests_config) > 1) and (not df_sgnf_results.empty):\n",
    "    query_test: str = 'M' if Selected_test_config_name == 'MCT-MCt,mCT-mCt' else 'T'\n",
    "    if (d := ut.find_XX_interesting_results(df_sgnf_results, query_test, debug=True)) is not None:\n",
    "        print(f\"\\nFound interesting {query_test}-{query_test} tests (with no matching tests):\")\n",
    "        display(d)\n",
    "\n",
    "        if save_interesting_results:\n",
    "            myut.dfs_to_excel_file([d], str(f\"{save_result_base_name}_in1not2.xlsx\"), ['Data'])\n",
    "    else:\n",
    "        print(\"No interesting results.\")\n",
    "\n",
    "else:\n",
    "    print('Not relevant.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" \n",
    "Save results to pickle and excel (df_sgnf_results is saved to excel)\n",
    "\"\"\"\n",
    "with open(f\"{save_result_base_name}.pkl\", 'wb') as fp:\n",
    "    pickle.dump(gene_results, fp)\n",
    "myut.dfs_to_excel_file([df_sgnf_results], str(f\"{str(save_result_base_name)}.xlsx\"), ['Data'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot a Result\n",
    "\n",
    "See the section `Plot All Results to PDF` to plot all significant results to a pdf file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" \n",
    "Load results if needed.\n",
    "\n",
    "THIS WILL OVERWRITE THE RESULTS ABOVE\n",
    "\"\"\"\n",
    "# #load_results_file_name: pathlib.Path = results_path / 'DSA_sgnf_splice_list_10_15_111-011_v2.pkl'\n",
    "# #load_results_file_name: pathlib.Path = results_path / 'DSA_sgnf_splice_list_10_15_111-011_v2.pkl'\n",
    "# load_results_file_name: pathlib.Path = pathlib.Path(f\"{save_result_base_name}.pkl\")\n",
    "# #load_results_file_name: pathlib.Path = results_path / 'DSA_sgnf_splice_list_10_15_111-011_110-010_v2.pkl'\n",
    "# # =======================================================================================================\n",
    "# print(f\"OVERWRITTING RESULTS WITH results from {load_results_file_name} !!\")\n",
    "# with open(load_results_file_name, 'rb') as fp:\n",
    "#     gene_results = pickle.load(fp)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" \n",
    "Show significant results. \n",
    "\n",
    "You can copy the candidate from this print and set it in the 'result_key' parameter in the next cell.\n",
    "\"\"\"\n",
    "for gn, gn_v in gene_results.items():\n",
    "    for k, v in gn_v.items():\n",
    "        print(f\"\\n['{gn}', '{k}']\")\n",
    "        for kv, vv in v.items():\n",
    "            if 'group' not in kv:\n",
    "                print(f\"\\t{kv}: {vv}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#result_key = ['TPX2', 'MCT-mCT|TPX2:chr20:Splice_Region:31770476:G:T|TCGA_LAML|ALL_patients']\n",
    "#result_key = ['TPX2', 'mCT-mCt|TPX2:chr20:Splice_Region:31770476:G:T|TCGA_BRCA|immunotherapy']\n",
    "#result_key = ['TPX2', 'McT-mcT|TPX2:chr20:Splice_Region:31770476:G:T|PAN-cancer|ALL_patients']\n",
    "#result_key = ['MET', 'McT-mcT|MET:chr7:Splice_Site:116739948:A:T|PAN-cancer|chemotherapy']\n",
    "#result_key = ['TPX2', 'MCT-mCT|TPX2:chr20:Splice_Region:31770476:G:T|TCGA_LAML|ALL_patients']\n",
    "\n",
    "\n",
    "#result_key = ['PIK3CA', 'MCT-mCT|MCT-mCT|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|hormone_therapy']\n",
    "#result_key = ['PIK3CA', 'MCT-mCT|MCT-mCT|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|tamoxifen']\n",
    "#result_key = ['PIK3CA', 'MCT-mCT|MCT-mCT|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|cyclophosphamide']\n",
    "#result_key = ['TPX2', 'MCT-mCT|MCT-mCT|TPX2:chr20:Splice_Region:31770476:G:T|TCGA_OV|taxol']\n",
    "#result_key = ['APBA2', 'MCT-mCT|APBA2:chr15:Splice_Site:29076093:T:G|TCGA_OV|taxol']\n",
    "#result_key = ['GLG1', 'MCT-mCT|GLG1:chr16:Splice_Site:74483125:C:T|TCGA_BRCA|doxorubicin']\n",
    "#result_key = ['MYOF', 'MCT-mCT|MYOF:chr10:Splice_Site:93379861:A:C|TCGA_OV|cisplatin']\n",
    "\n",
    "#result_key = ['RAD51D', 'MCT-mCT|MCT-mCT|RAD51D:chr17:Splice_Site:35118499:A:C|TCGA_GBM|temodar']\n",
    "#result_key = ['HERC3', 'MCT-mCT|MCT-mCT|HERC3:chr4:Splice_Site:88655306:T:G|TCGA_OV|doxil']\n",
    "#result_key = ['CAD', 'MCT-mCT|MCT-mCT|CAD:chr2:Splice_Site:27218017:G:T|TCGA_BRCA|hormone_therapy']\n",
    "#result_key = ['CDON', 'MCT-mCT|MCT-mCT|CDON:chr11:Splice_Site:126021246:A:C|TCGA_OV|carboplatin']\n",
    "\n",
    "#result_key = ['ATP1A2', 'MCT-mCT|MCT-mCT|ATP1A2:chr1:Splice_Site:160135297:T:G|TCGA_GBM|dexamethasone']\n",
    "#result_key = ['DYRK4', 'MCT-mCT|MCT-mCT|DYRK4:chr12:Splice_Site:4591300:T:G|TCGA_BRCA|arimidex']\n",
    "#result_key = ['ANKRD17', 'MCT-mCT|MCT-mCT|ANKRD17:chr4:Splice_Site:73139529:A:C|TCGA_OV|topotecan']\n",
    "#result_key = ['MSH5', 'MCT-mCT|MCT-mCT|MSH5:chr6:Splice_Site:31759977:T:G|TCGA_GBM|chemotherapy']\n",
    "#result_key = ['EMC1', 'MCT-mCT|MCT-mCT|EMC1:chr1:Splice_Site:19243612:A:C|TCGA_LUAD|chemotherapy']\n",
    "#result_key = ['MTMR11', 'MCT-mCT|MCT-mCT|MTMR11:chr1:Splice_Site:149935297:A:C|TCGA_OV|taxol']\n",
    "\n",
    "#result_key = ['PIK3CA', 'MCT-mCT|MCT-mCT|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|taxol']\n",
    "#result_key = ['PIK3CA', 'MCT-mCT,MCt-mCt|MCT-mCT|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|arimidex']\n",
    "\n",
    "#result_key = ['MTMR11', 'MCT-mCT,MCt-mCt|MCT-mCT|MTMR11:chr1:Splice_Site:149935297:A:C|TCGA_BRCA|doxorubicin']\n",
    "\n",
    "#result_key = ['POLR3B', 'MCT-MCt,mCT-mCt|MCT-MCt|POLR3B:chr12:Splice_Site:106369685:T:G|TCGA_BRCA|docetaxel']\n",
    "#result_key = ['SPEG', 'MCT-MCt,mCT-mCt|MCT-MCt|SPEG:chr2:Splice_Site:219477787:T:G|TCGA_PAAD|chemotherapy']\n",
    "#result_key = ['SPEG', 'MCT-MCt,mCT-mCt|MCT-MCt|SPEG:chr2:Splice_Site:219477787:T:G|TCGA_PAAD|gemcitabine']\n",
    "\n",
    "#result_key = ['SPEG', 'MCT-mCT|SPEG:chr2:Splice_Site:219477787:T:G|TCGA_PAAD|ALL_patients']\n",
    "\n",
    "#result_key = ['AASS', 'MCT-MCt,mCT-mCt|MCT-MCt|AASS:chr7:Splice_Site:122126373:A:C|TCGA_GBM|dexamethasone']\n",
    "#result_key = ['ATP1A2', 'MCT-MCt,mCT-mCt|MCT-MCt|ATP1A2:chr1:Splice_Site:160135297:T:G|TCGA_GBM|dexamethasone']\n",
    "#result_key = ['PRRC2A', 'MCT-MCt,mCT-mCt|MCT-MCt|PRRC2A:chr6:Splice_Site:31629336:T:G|TCGA_GBM|dexamethasone']\n",
    "#result_key = ['RAD51D', 'MCT-MCt,mCT-mCt|MCT-MCt|RAD51D:chr17:Splice_Site:35118499:A:C|TCGA_GBM|dexamethasone']\n",
    "#result_key = ['TSHR', 'MCT-MCt,mCT-mCt|MCT-MCt|TSHR:chr14:Intron:81096719:G:T|TCGA_GBM|dexamethasone']\n",
    "#result_key = ['TSHR', 'MCT-MCt,mCT-mCt|MCT-MCt|TSHR:chr14:Intron:81096719:G:T|TCGA_LUSC|chemotherapy']\n",
    "#result_key = ['SPTBN1', 'MCT-MCt,mCT-mCt|MCT-MCt|SPTBN1:chr2:Splice_Site:54599245:T:G|TCGA_GBM|dexamethasone']\n",
    "\n",
    "#result_key = ['MET', 'MCT-mCT|MCT-mCT|MET:chr7:Intron:116769622:TTT:-|TCGA_GBM|chemotherapy']\n",
    "#result_key = ['PIK3CA', 'MCT-mCT|MCT-mCT|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|doxorubicin']\n",
    "\n",
    "#result_key = ['EGFR', 'MCT-MCt,mCT-mCt|MCT-MCt|EGFR:chr7:Intron:55201370:TC:-|TCGA_LGG|bevacizumab']\n",
    "\n",
    "#result_key = ['PIK3CA', 'MCT-mCT,MCt-mCt|MCt-mCt|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|hormone_therapy']\n",
    "#result_key = ['PIK3CA', 'MCT-mCT,MCt-mCt|MCT-mCT|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|tamoxifen']\n",
    "#result_key = ['PIK3CA', 'MCT-mCT,MCt-mCt|MCt-mCt|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|tamoxifen']\n",
    "\n",
    "#result_key = ['PIK3CA', 'MCT-mCT,MCt-mCt|MCT-mCT|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|taxol']\n",
    "#result_key = ['PIK3CA', 'MCT-mCT,MCt-mCt|MCt-mCt|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|taxol']\n",
    "\n",
    "#result_key = ['TPX2', 'MCT-mCT,MCt-mCt|MCt-mCt|TPX2:chr20:Splice_Region:31770476:G:T|TCGA_OV|taxol']\n",
    "#result_key = ['GLG1', 'MCT-mCT,MCt-mCt|MCt-mCt|GLG1:chr16:Splice_Site:74483125:C:T|TCGA_BRCA|doxorubicin']\n",
    "#result_key = ['MTMR11', 'MCT-mCT,MCt-mCt|MCt-mCt|MTMR11:chr1:Splice_Site:149935297:A:C|TCGA_BRCA|doxorubicin']\n",
    "#result_key = ['HERC3', 'MCT-mCT,MCt-mCt|MCt-mCt|HERC3:chr4:Splice_Site:88655306:T:G|TCGA_OV|doxil']\n",
    "#result_key = ['CAD', 'MCT-mCT,MCt-mCt|MCt-mCt|CAD:chr2:Splice_Site:27218017:G:T|TCGA_BRCA|hormone_therapy']\n",
    "#result_key = ['ATP1A2', 'MCT-mCT,MCt-mCt|MCt-mCt|ATP1A2:chr1:Splice_Site:160135297:T:G|TCGA_GBM|dexamethasone']\n",
    "#result_key = ['DYRK4', 'MCT-mCT,MCt-mCt|MCt-mCt|DYRK4:chr12:Splice_Site:4591300:T:G|TCGA_BRCA|arimidex']\n",
    "#result_key = ['MSH5', 'MCT-mCT,MCt-mCt|MCT-mCT|MSH5:chr6:Splice_Site:31759977:T:G|TCGA_GBM|chemotherapy']\n",
    "#result_key = ['EMC1', 'MCT-mCT,MCt-mCt|MCt-mCt|EMC1:chr1:Splice_Site:19243612:A:C|TCGA_LUAD|chemotherapy']\n",
    "#result_key = ['RAD51D', 'MCT-mCT,MCt-mCt|MCt-mCt|RAD51D:chr17:Splice_Site:35118499:A:C|TCGA_GBM|temodar']\n",
    "#result_key = ['CDON', 'MCT-mCT,MCt-mCt|MCT-mCT|CDON:chr11:Splice_Site:126021246:A:C|TCGA_OV|carboplatin']\n",
    "#result_key = ['EGFR', 'MCT-MCt,mCT-mCt|mCT-mCt|EGFR:chr7:Intron:55201370:TC:-|TCGA_LGG|bevacizumab']\n",
    "#result_key = ['POLR3B', 'MCT-MCt,mCT-mCt|mCT-mCt|POLR3B:chr12:Splice_Site:106369685:T:G|TCGA_BRCA|docetaxel']\n",
    "#result_key = ['SPEG', 'MCT-MCt,mCT-mCt|mCT-mCt|SPEG:chr2:Splice_Site:219477787:T:G|TCGA_PAAD|chemotherapy']\n",
    "#result_key = ['SPEG', 'MCT-MCt,mCT-mCt|mCT-mCt|SPEG:chr2:Splice_Site:219477787:T:G|TCGA_PAAD|gemcitabine']\n",
    "#result_key = ['ATP1A2', 'MCT-MCt,mCT-mCt|mCT-mCt|ATP1A2:chr1:Splice_Site:160135297:T:G|TCGA_GBM|dexamethasone']\n",
    "#result_key = ['PRRC2A', 'MCT-MCt,mCT-mCt|mCT-mCt|PRRC2A:chr6:Splice_Site:31629336:T:G|TCGA_GBM|dexamethasone']\n",
    "#result_key = ['AASS', 'MCT-MCt,mCT-mCt|mCT-mCt|AASS:chr7:Splice_Site:122126373:A:C|TCGA_GBM|dexamethasone']\n",
    "#result_key = ['RAD51D', 'MCT-MCt,mCT-mCt|mCT-mCt|RAD51D:chr17:Splice_Site:35118499:A:C|TCGA_GBM|dexamethasone']\n",
    "#result_key = ['TSHR', 'MCT-MCt,mCT-mCt|mCT-mCt|TSHR:chr14:Intron:81096719:G:T|TCGA_GBM|dexamethasone']\n",
    "#result_key = ['TSHR', 'MCT-MCt,mCT-mCt|mCT-mCt|TSHR:chr14:Intron:81096719:G:T|TCGA_LUSC|chemotherapy']\n",
    "#result_key = ['SPTBN1', 'MCT-MCt,mCT-mCt|mCT-mCt|SPTBN1:chr2:Splice_Site:54599245:T:G|TCGA_GBM|dexamethasone']\n",
    "\n",
    "#result_key = ['BCL9L', 'MCT-mCT,MCt-mCt|MCT-mCT|BCL9L:chr11:Intron:118903196:AG:-|TCGA_OV|paclitaxel']\n",
    "#result_key = ['BCL9L', 'MCT-mCT,MCt-mCt|MCT-mCT|BCL9L:chr11:Intron:118903196:AG:-|TCGA_OV|carboplatin']\n",
    "#result_key = ['BCL9L', 'MCT-mCT,MCt-mCt|MCt-mCt|BCL9L:chr11:Intron:118903196:AG:-|TCGA_OV|carboplatin']\n",
    "#result_key = ['BCL9L', 'MCT-mCT,MCt-mCt|MCt-mCt|BCL9L:chr11:Intron:118903196:AG:-|TCGA_OV|paclitaxel']\n",
    "\n",
    "# not significant results (we set p threshold to 0.99)\n",
    "#result_key = ['PIK3CA', 'MCT-mCT,MCt-mCt|MCT-mCT|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|adriamycin']\n",
    "#result_key = ['PIK3CA', 'MCT-mCT,MCt-mCt|MCt-mCt|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|adriamycin']\n",
    "#result_key = ['PIK3CA', 'MCT-mCT,MCt-mCt|MCt-mCt|PIK3CA:chr3:Intron:179203524:G:C|TCGA_BRCA|cytoxan']\n",
    "\n",
    "#result_key = ['BCL9L', 'MCT-mCT,MCt-mCt|MCT-mCT|BCL9L:chr11:Intron:118903196:AG:-|TCGA_OV|chemotherapy']\n",
    "#result_key = ['BCL9L', 'MCT-mCT,MCt-mCt|MCT-mCT|BCL9L:chr11:Intron:118903196:AG:-|TCGA_OV|taxol']\n",
    "\n",
    "#result_key = ['TPX2', 'MCT-mCT,MCt-mCt|MCT-mCT|TPX2:chr20:Splice_Region:31770476:G:T|TCGA_OV|carboplatin']\n",
    "#result_key = ['TPX2', 'MCT-mCT,MCt-mCt|MCT-mCT|TPX2:chr20:Splice_Region:31770476:G:T|TCGA_OV|chemotherapy']\n",
    "\n",
    "#result_key = ['GLG1', 'MCT-mCT,MCt-mCt|MCT-mCT|GLG1:chr16:Splice_Site:74483125:C:T|TCGA_BRCA|cytoxan']  # cyclophosphamide  cytoxan\n",
    "#result_key = ['GLG1', 'MCT-mCT,MCt-mCt|MCt-mCt|GLG1:chr16:Splice_Site:74483125:C:T|TCGA_BRCA|cytoxan']\n",
    "\n",
    "#result_key = ['APBA2', 'MCT-mCT,MCt-mCt|MCT-mCT|APBA2:chr15:Splice_Site:29076093:T:G|TCGA_OV|cisplatin']\n",
    "\n",
    "#result_key = ['MYOF', 'MCT-mCT,MCt-mCt|MCT-mCT|MYOF:chr10:Splice_Site:93379861:A:C|TCGA_OV|carboplatin']\n",
    "\n",
    "#result_key = ['MTMR11', 'MCT-mCT,MCt-mCt|MCT-mCT|MTMR11:chr1:Splice_Site:149935297:A:C|TCGA_BRCA|adriamycin']\n",
    "\n",
    "#result_key = ['HERC3', 'MCT-mCT,MCt-mCt|MCT-mCT|HERC3:chr4:Splice_Site:88655306:T:G|TCGA_OV|carboplatin']\n",
    "\n",
    "#result_key = ['CAD', 'MCT-mCT,MCt-mCt|MCt-mCt|CAD:chr2:Splice_Site:27218017:G:T|TCGA_BRCA|adriamycin']\n",
    "\n",
    "# result_key = ['ATP1A2', 'MCT-mCT,MCt-mCt|MCT-mCT|ATP1A2:chr1:Splice_Site:160135297:T:G|TCGA_GBM|temodar']\n",
    "\n",
    "#result_key = ['DYRK4', 'MCT-mCT,MCt-mCt|MCt-mCt|DYRK4:chr12:Splice_Site:4591300:T:G|TCGA_BRCA|cytoxan']\n",
    "\n",
    "#result_key = ['RAD51D', 'MCT-mCT,MCt-mCt|MCt-mCt|RAD51D:chr17:Splice_Site:35118499:A:C|TCGA_GBM|temozolomide']\n",
    "\n",
    "#result_key = ['CDON', 'MCT-mCT,MCt-mCt|MCT-mCT|CDON:chr11:Splice_Site:126021246:A:C|TCGA_OV|chemotherapy']\n",
    "#result_key = ['CDON', 'MCT-mCT,MCt-mCt|MCT-mCT|CDON:chr11:Splice_Site:126021246:A:C|TCGA_OV|taxol']\n",
    "\n",
    "#result_key = ['EGFR', 'MCT-MCt,mCT-mCt|mCT-mCt|EGFR:chr7:Intron:55201370:TC:-|TCGA_LGG|lomustine']\n",
    "\n",
    "#result_key = ['TPX2', 'MCT-mCT|MCT-mCT|TPX2:chr20:Splice_Region:31770476:G:T|TCGA_LAML|ALL_patients']\n",
    "#result_key = ['MTMR11', 'MCT-mCT|MCT-mCT|MTMR11:chr1:Splice_Site:149935297:A:C|TCGA_LAML|ALL_patients']\n",
    "#result_key = ['GSN', 'MCT-mCT|MCT-mCT|GSN:chr9:Splice_Site:121300007:T:G|TCGA_LIHC|ALL_patients']\n",
    "#result_key = ['SPEG', 'MCT-mCT|MCT-mCT|SPEG:chr2:Splice_Site:219477787:T:G|TCGA_PAAD|ALL_patients']\n",
    "#result_key = ['HERC3', 'MCT-mCT|MCT-mCT|HERC3:chr4:Splice_Site:88655306:T:G|TCGA_GBM|ALL_patients']\n",
    "#result_key = ['CHL1', 'MCT-mCT|MCT-mCT|CHL1:chr3:Splice_Site:401624:A:T|TCGA_LAML|ALL_patients']\n",
    "#result_key = ['APBA2', 'MCT-mCT|MCT-mCT|APBA2:chr15:Splice_Site:29076093:T:G|TCGA_LUSC|ALL_patients']\n",
    "\n",
    "#result_key = ['TRPC4AP', 'MCT-mCT|MCT-mCT|TRPC4AP:chr20:Splice_Region:35054969:T:G|TCGA_KIRC|ALL_patients']\n",
    "\n",
    "#result_key = ['GSN', 'MCT-mCT|MCT-mCT|GSN:chr9:Splice_Site:121300007:T:G|TCGA_PAAD|ALL_patients']\n",
    "\n",
    "#result_key = ['APBA2', 'MCT-mCT|MCT-mCT|APBA2:chr15:Splice_Site:29076093:T:G|TCGA_CESC|ALL_patients']\n",
    "#result_key = ['NBPF3', 'MCT-mCT|MCT-mCT|NBPF3:chr1:Splice_Site:21473586:G:T|TCGA_STAD|ALL_patients']\n",
    "#result_key = ['RAD51D', 'MCT-mCT|MCT-mCT|RAD51D:chr17:Splice_Site:35118499:A:C|TCGA_LAML|ALL_patients']\n",
    "result_key = ['HERC3', 'MCT-mCT|MCT-mCT|HERC3:chr4:Splice_Site:88655306:T:G|TCGA_CESC|ALL_patients']\n",
    "\n",
    "\n",
    "\n",
    "figsize = (12,6)\n",
    "linewidth: int = 3\n",
    "fontsize: int = 14\n",
    "group1_color, group2_color = 'red', 'blue'\n",
    "group1_lbl, group2_lbl = result_key[1].split('|')[1].split('-')\n",
    "p_value_txt_pos = (0.4, 0.75)\n",
    "\n",
    "save_fig = False\n",
    "# RECALL that the OS file names is case insensitive\n",
    "save_fig_file = report_path / f\"survival__{result_key[1].replace(':', '_').replace('-', 'X').replace('|', '__').replace(',', '_')}.png\"\n",
    "# ====================================================================================================================================\n",
    "gene, r_key = result_key\n",
    "#title: str = f\"KM curves for {','.join(r_key.split('|')[1:])}\"\n",
    "#title: str = f\"KM curves for {','.join(r_key.split('|'))}\"\n",
    "title: str = f\"KM curves for {r_key}\"\n",
    "\n",
    "fig, ax = ut.plot_gene_muts_survival_result(gene_results[gene][r_key], title,\n",
    "                                              group1_lbl, group2_lbl,\n",
    "                                              group1_color=group1_color, group2_color=group2_color,\n",
    "                                              figsize=figsize, p_value_txt_pos=p_value_txt_pos,\n",
    "                                              linewidth=linewidth, fontsize=fontsize)\n",
    "\n",
    "if save_fig:\n",
    "    plt.savefig(save_fig_file)\n",
    "    print(f\"\\nFigure saved in {save_fig_file}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot All Results' Curves to PDF\n",
    "\n",
    "We plot all significant results to a pdf file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf_file = pathlib.Path(f\"{save_result_base_name}_all_figs.pdf\")\n",
    "exclude_genes: tuple[str] = () #('TPX2', 'MTMR11',)  # results of these genes will be skipped\n",
    "figsize: tuple = (12, 6,)\n",
    "generate_title_page: bool = True\n",
    "title_page_text: str = ''  # if set to '', the code will create a general title with date and time.\n",
    "# ======================================================================================================\n",
    "with plt.ioff():\n",
    "    ut.plot_results_KM_curves_to_pdf(pdf_file, gene_results, exclude_genes=exclude_genes, figsize=figsize, generate_title_page=generate_title_page, title_page_text=title_page_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extra"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clinical Data with Drugs\n",
    "In this section we prepare the variable `df_clinical_drugs_dict` that is used in the `Initialization` section above. We do it here to save time on multiple initializations.\n",
    "\n",
    "Thus, **no** need to run this section when performing DSA above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import data_utils as dtut\n",
    "import GDC_dataset_feb2021_utils as gdcu\n",
    "\n",
    "min_drug_prevalence: int = 20  #50 #100  # analyzing only drugs used by at least this number of patients\n",
    "save_file_name: pathlib.Path = report_path / 'TCGA_clinical_preprocess' / f'df_clinical_drugs_dict_{min_drug_prevalence}.pkl'\n",
    "save_combined_file_name: pathlib.Path = report_path / 'TCGA_clinical_preprocess' / f'df_clinical_therapies_drugs_dict_{min_drug_prevalence}.pkl'\n",
    "\n",
    "#Event_col_label: str = 'event'\n",
    "#Duration_col_label: str = 'duration'\n",
    "\n",
    "# keys for clinical data of all patients, all patients with no therapy and all patients with a therapy, respectively\n",
    "#All_name: str = 'ALL_patients'\n",
    "#None_therapy_name: str = 'NO_therapy'\n",
    "#All_therapies_name: str = 'ALL_therapies'\n",
    "# ===========================================================================================================================\n",
    "\n",
    "# load TCGA clinical data\n",
    "clinical_data_file: pathlib.Path = pathlib.Path(si.ROOT_path) / 'Data' / 'GDC' / 'Clinical' / 'df_p_all.pkl'\n",
    "df_clinical = pd.read_pickle(clinical_data_file)\n",
    "\n",
    "# add event and duration columns (part of this function)\n",
    "df_clinical_therapies_dict = dtut.prepare_clinical_dfs(df_clinical, gdcu.Therapies, Dsa_data.event_name, Dsa_data.duration_name, \n",
    "                                                       clinical_case_id_col=Dsa_data.patient_ID_col, \n",
    "                                                       all_name=Dsa_data.All_name, none_therapy_name=Dsa_data.None_therapy_name, all_therapies_name=Dsa_data.All_therapies_name)\n",
    "\n",
    "# generate clinical data per drug\n",
    "df_clinical_drugs_dict, _ = gdcu.get_drugs_clinical_dfs(df_clinical_therapies_dict[Dsa_data.All_name], min_drug_prevalence=min_drug_prevalence)\n",
    "\n",
    "# saving only the drug data\n",
    "with open(save_file_name, 'wb') as fp:\n",
    "    pickle.dump(df_clinical_drugs_dict, fp)\n",
    "print(f\"{save_file_name} created.\")\n",
    "\n",
    "# saving combined therapies and drugs data\n",
    "with open(save_combined_file_name, 'wb') as fp:\n",
    "    pickle.dump(df_clinical_therapies_dict | df_clinical_drugs_dict, fp)\n",
    "print(f\"{save_combined_file_name} created.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ScratchPad"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TOML configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import DSA_config\n",
    "import toml_utils as tmut  # just to show the data\n",
    "\n",
    "print(DSA_config.data)\n",
    "tmut.print_nested_dicts(DSA_config.data)\n",
    "\n",
    "print(DSA_config.clinical_data_file)\n",
    "#display(DSA_config.df_clinical)\n",
    "\n",
    "for k, v in DSA_config.df_clinical_therapies_dict.items():\n",
    "    print(k, v.shape[0])\n",
    "\n",
    "print(DSA_config.Logrank_kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import toml_utils as tmut\n",
    "\n",
    "toml_file = pathlib.Path('./DSA_config/config.toml')\n",
    "# =================================================\n",
    "toml_data = tmut.myToml().load(toml_file)\n",
    "print(toml_data)\n",
    "tmut.print_nested_dicts(toml_data)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Concat multiple batch results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" \n",
    "Concatenating different batch results (both excel and pickle)\n",
    "\"\"\"\n",
    "batch_indexes: list = list(range(1, 4+1))\n",
    "base_file_template: str = str(results_path / f\"survival_sgnf_splice_list_10_15_M_P_w_wo_batch_?_4\")\n",
    "\n",
    "save_concat: bool = True\n",
    "save_concat_base_name = base_file_template.split('_batch')[0]\n",
    "# =================================================================================================\n",
    "concat_df, concat_d, return_ok_code = ut.concat_batch_DS_result_files(batch_indexes, base_file_template, index_place_holder_str='?')\n",
    "if return_ok_code:\n",
    "    display(concat_df)\n",
    "    if save_concat:\n",
    "        with open(f\"{save_concat_base_name}.pkl\", 'wb') as fp:\n",
    "            pickle.dump(concat_d, fp)\n",
    "        myut.dfs_to_excel_file([concat_df], save_concat_base_name + '.xlsx', ['Data'])\n",
    "else:\n",
    "    print(f\"Mismatch between number of cases in concat_df and concat_d !!\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv311",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
